webapp/src/serviceworker.ts

```diff
@@ -217,3 +224,184 @@ function initWebappServiceWorker() {
+
+                if (message.didDisconnect) lockGranted = undefined;
+                if (pendingDisconnectResolver) pendingDisconnectResolver(message.didDisconnect ? DisconnectResponse.Disconnected : DisconnectResponse.Waiting);
+            }
+            else if (message.action === "packet-io-supported") {
+                await sendToAllClientsAsync({
+                    type: "serviceworker",
+                    action: "packet-io-supported",
+                    supported: true
+                });
+            }
+            else if (message.action === "packet-io-status" && message.hasLock && statusResolver) {
+                statusResolver(message.lock);
+            }
+        }
+    });
+
+    async function sendToAllClientsAsync(message: pxt.ServiceWorkerMessage) {
+        const clients = await (self as unknown as ServiceWorkerScope).clients.matchAll()
+        clients.forEach(c => c.postMessage(message))
+    }
+
+    // Waits for the disconnect and times-out after 5 seconds if there is no response
+    function waitForLockDisconnectAsync() {
+        let ref: any;
+
+        const promise = new Promise<DisconnectResponse>((resolve) => {
+            console.log("Waiting for disconnect " + lockGranted);
+            pendingDisconnectResolver = resolve;
+            sendToAllClientsAsync({
+                type: "serviceworker",
+                action: "packet-io-lock-disconnect",
+                lock: lockGranted
+            });
+        })
+
+        const timeoutPromise = new Promise<DisconnectResponse>(resolve => {
+            ref = setTimeout(() => {
+                console.log("Timed-out disconnect request " + lockGranted);
+                resolve(DisconnectResponse.TimedOut);
+            }, 5000)
+        });
+
+        return Promise.race([ promise, timeoutPromise ])
+            .then(resp => {
+                clearTimeout(ref);
+                pendingDisconnectResolver = undefined
+                return resp;
+            });
+    }
+
+    function checkForExistingLockAsync() {
+        if (lockGranted) return Promise.resolve(lockGranted);
+        let ref: any;
+
+        const promise = new Promise<string>(resolve => {
+            console.log("check for existing lock");
+            statusResolver = resolve;
+            sendToAllClientsAsync({
+                type: "serviceworker",
+                action: "packet-io-status"
+            });
+        })
+
+        const timeoutPromise = new Promise<string>(resolve => {
+            ref = setTimeout(() => {
+                console.log("Timed-out disconnect request " + lockGranted);
+                resolve(undefined);
+            }, 1000)
+        });
+
+        return Promise.race([ promise, timeoutPromise ])
+            .then(resp => {
+                clearTimeout(ref);
+                statusResolver = undefined
+                return resp;
+            });
+    }
+
+
+    function delay(millis: number): Promise<void> {
```

Comment:
Just to confirm, not using the helpers in `pxtlib/util.ts` for things like delay / timeout here because serviceworker file is run separate from the rest / don't want any deps on other files?

Reply:
correct!